packages/yoastseo/src/parse/build/private/getSentencePositions.js

```diff
@@ -0,0 +1,120 @@
+ *
+ */
+function getDescendantPositions( descendantNodes ) {
+	const descendantTagPositions = [];
+	descendantNodes.forEach( ( node ) => {
+		descendantTagPositions.push( node.sourceCodeLocation.startTag );
+		descendantTagPositions.push( node.sourceCodeLocation.endTag );
+	} );
+	// Sort the tag position objects by the start tag position in ascending order.
+	descendantTagPositions.sort( ( a, b ) => a.startOffset - b.startOffset );
+
+	return descendantTagPositions;
+}
+
+/**
+ *
+ * Adjusts the end position of the sentence to account for any descendant node tags that overlap with the sentence.
+ * We want to add the length of any tags that are within the sentence, as well as start tags at the beginning of the
+ * sentence and end tags at the end of the sentence, to the end position of the sentence.
+ * We don't want to include start tags at the end of the sentence in order to reduce the risk of invalid HTML after
+ * adding mark tags around the sentence. (Same goes for end tags at the start of the sentence, but this should already
+ * never happen, as these tags should be added to the end position of the previous sentence).
+ *
+ * @param {Node[]}				descendantNodes			The descendant nodes.
+ * @param {SourceCodeRange[]}	descendantTagPositions	The positions of the descendant nodes' tags.
+ * @param {number}				sentenceStart			The start position of a sentence.
+ * @param {number}				sentenceEnd				The end position of a sentence.
+ *
+ * @returns {number}	The adjusted end position of the sentence.
+ */
+function adjustSentenceEnd( descendantNodes, descendantTagPositions, sentenceStart, sentenceEnd ) {
+	/*
+	 * If the start position of a descendant's node tag is between the start and end position of the sentence, or is
+	 * the same as the start/end position of the sentence, add the tag's length to the end position of the sentence.
+	 */
+	descendantTagPositions.forEach( ( position ) => {
+		if ( position.startOffset >= sentenceStart && position.startOffset <= sentenceEnd ) {
+			sentenceEnd += ( position.endOffset - position.startOffset );
+		}
+	} );
+	/*
+	 * If the length of a start tag at the end of the sentence was added to the sentenceEnd in the step above,
+	 * remove it. We are using the data from the original descendantNodes array for this check, as the
+	 * descendantTagPositions array doesn't specify whether each tag is an opening or closing tag.
+	 */
+	const startTagAtEndOfSentence = descendantNodes.find( node =>
+		node.sourceCodeLocation.startTag.endOffset === sentenceEnd );
+	if ( startTagAtEndOfSentence ) {
+		const startTagLocation = startTagAtEndOfSentence.sourceCodeLocation.startTag;
+		sentenceEnd = sentenceEnd - ( startTagLocation.endOffset - startTagLocation.startOffset );
+	}
+
+	return sentenceEnd;
+}
+
+/**
+ * Gets the start and end positions of the sentences in the source code.
+ *
+ * @param {Node} 		node  		The paragraph or heading node.
```

Comment:
If you're clear on always passing a `Paragraph` or `Heading`, you could make that explicit in the type definition (`{Paragraph|Heading}`). But in this case let's not do that.

Reply:
You mean we shouldn't that because we might want to also use the function for other nodes? In that case I guess I should also change the description of the parameter from 'The paragraph or heading node' to something more general.

Reply:
Hmm, on second thought, let's do change the type definition here to align with the description, but then also change it when you call the function in `packages/yoastseo/src/parse/build/private/tokenize.js`.